naps/cores/cmv12k/pixel_remapper.py

```diff
@@ -4,56 +4,334 @@
+# The sensor output is sturctured into bursts of 128 pixels, so we need to pay attention to the burst boundaries, as after a burst some overhead cycles take place
+# channel written:
+# | clock -> | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
+# |   bram ∨ |   |   |   |   |   |   |   |   |
+# |----------+---+---+---+---+---+---+---+---|
+# |        0 | 0 | 2 | 4 | 6 | 1 | 3 | 5 | 6 |
+# |        1 | 1 | 3 | 5 | 7 | 0 | 2 | 4 | 7 |
+# the addresses written are chosen to just simplify the read out sequence:
+# address:
+# | clock -> |  1 |   2 |   3 |   4 |  5 |   6 |   7 |   8 |
+# |   bram ∨ |    |     |     |     |    |     |     |     |
+# |----------+----+-----+-----+-----+----+-----+-----+-----+
+# |        0 |  0 | 128 | 256 | 384 | 64 | 172 | 320 | 448 |
+# |        1 | 64 | 172 | 320 | 448 |  0 | 128 | 256 | 384 |
+#
+# on the read out side, we finally can just read the words sequentially out of the BRAMs
+# reading:
+# | clock -> | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
+# |   bram ∨ |   |   |   |   |   |   |   |   |
+# |----------+---+---+---+---+---+---+---+---|
+# |        0 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+# |        1 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+# a final reordering has to take place, as the first BRAM written to rotates through wrt to the channel
+# output order (the number in the table gives the BRAM the word was read from)
+# |  clock -> | 1 | 64 | 128 | 172 | 256 |
+# | channel ∨ |   |    |     |     |     |
+# |-----------+---+----+-----+-----+-----|
+# |         0 | 0 |  1 |   0 |   1 |   0 |
+# |         1 | 0 |  1 |   0 |   1 |   0 |
+# |         2 | 1 |  0 |   1 |   0 |   1 |
+# |         3 | 1 |  0 |   1 |   0 |   1 |
+#
+# TODO(robin): can we avoid that barrel shifter somehow?, Maybe a pipelined shifter?
+#
+# Now lets finally talk about thae handling of top and bottom side. We know that the bottom side transmits the row immediately after the row of the row transmitted by the top side.
+# So if we order the channels to be in the [top_channel, bottom_channel] order, we can just treat this as one row of twice the width (8196) and everything still works.
+#
+# Finally, all thats left is some scheduling logic, that start the readout process once enough data has be written to the BRAMs.
+# n_bram * 2 pixels are read each clock cycle.
+# And a single channel gets 1 pixel per clock cycle, for a total of 4096 / n_channel pixels. For example for 8 channels (per side): 512 pixels.
+# This means we read out all of the 512 pixels in 32 cycles, so we can start the read out process with 32 pixels left.
+# The next row will start soon after the first, so we store two rows and ping pong the buffers
+# TODO(robin): this is fucked because of OH, no?
 class Cmv12kPixelRemapper(Elaboratable):
-    """Remaps one output port (if top and bottom outputs of the cmv12k are used, instantiate this twice!) to provide a linear output.
+    def __init__(self, input: ImageStream, n_bits=12, two_sided_readout=True):
```

Comment:
why is the input an image stream though?

Reply:
and are input and output domains the same?

Reply:
yes, input and output domains are the same.
This is a `ImageStream`, because the cmv12k rx core outputs an `ImageStream`, I just built something that plugs into this. 
But of course we could change this. We do not have a "clear" definition of  `ImageStream` currently and the rx core also output `line_last` and `frame_last`, so it kind of makes sense to have its output be an `ImageStream`, but of course the pixel ordering is a bit "weird".